core/storage/trie/polkadot_trie/polkadot_trie_cursor_impl.cpp

```diff
@@ -167,153 +188,192 @@ namespace kagome::storage::trie {
-    OUTCOME_TRY(seekLowerBoundInternal(trie_.getRoot(), nibbles));
+    SAFE_VOID_CALL(seekLowerBoundInternal(*trie_->getRoot(), nibbles))
     return outcome::success();
   }
 
-  outcome::result<bool>
-  PolkadotTrieCursorImpl::seekNodeWithValueBothDirections() {
-    while (not last_visited_child_.empty()) {
-      auto [parent, idx] = last_visited_child_.back();
-      last_visited_child_.pop_back();
-      auto parent_node = std::dynamic_pointer_cast<PolkadotNode>(parent);
-      OUTCOME_TRY(ok, setChildWithMinIdx(parent_node, idx + 1));
-      if (ok) {
-        OUTCOME_TRY(seekNodeWithValue(parent_node));
-        current_ = parent_node;
+  outcome::result<bool> PolkadotTrieCursorImpl::nextNodeWithValueInOuterTree() {
+    BOOST_ASSERT(std::holds_alternative<SearchState>(state_));
+    auto &search_state = std::get<SearchState>(state_);
+
+    const auto &search_path = search_state.getPath();
+    while (not search_path.empty()) {
+      auto [parent, idx] = search_path.back();
+      BOOST_VERIFY_MSG(search_state.leaveChild(),
+                       "Guaranteed by the loop condition");
+      SAFE_CALL(child, visitChildWithMinIdx(parent, idx + 1))
+      if (child != nullptr) {
+        SAFE_VOID_CALL(nextNodeWithValueInSubTree(*child))
         return true;
       }
     }
     return false;
   }
 
-  outcome::result<void> PolkadotTrieCursorImpl::seekNodeWithValue(
-      NodePtr &parent) {
-    if (parent->value.has_value()) {
-      return outcome::success();
-    }
-    if (not parent->isBranch()) {
-      return Error::INVALID_NODE_TYPE;
+  outcome::result<void> PolkadotTrieCursorImpl::nextNodeWithValueInSubTree(
+      const TrieNode &parent) {
+    auto *current = &parent;
+    while (not current->value.has_value()) {
+      if (not current->isBranch()) {
+        return Error::INVALID_NODE_TYPE;
+      }
+      SAFE_CALL(child, visitChildWithMinIdx(*current))
+      BOOST_ASSERT_MSG(child != nullptr, "Branch node must contain a leaf");
+      current = child;
     }
-    OUTCOME_TRY(setChildWithMinIdx(parent));
-    OUTCOME_TRY(seekNodeWithValue(parent));
     return outcome::success();
   }
 
   outcome::result<void> PolkadotTrieCursorImpl::seekUpperBound(
-      const common::Buffer &key) {
-    OUTCOME_TRY(seekLowerBound(key));
-    if (this->key().has_value() and this->key() == key) {
-      OUTCOME_TRY(next());
-    }
+      const common::BufferView &key) {
+    SAFE_VOID_CALL(seekLowerBound(key))
+    SAFE_VOID_CALL(next())
     return outcome::success();
   }
 
-  outcome::result<bool> PolkadotTrieCursorImpl::setChildWithMinIdx(
-      NodePtr &parent, uint8_t min_idx) {
-    auto branch = std::dynamic_pointer_cast<BranchNode>(parent);
+  outcome::result<const TrieNode *>
+  PolkadotTrieCursorImpl::visitChildWithMinIdx(const TrieNode &parent,
+                                               uint8_t min_idx) {
+    BOOST_ASSERT(std::holds_alternative<SearchState>(state_));
```

Comment:
Yes, but given that the next thing we do is extract SearchState I think it's clearer to leave it this way, because isValid's connection to Search state is not immediately recognisable.